### PR TITLE
New api

### PR DESCRIPTION
**Pull request by ericdevries**
Fixes DD-1341 and DD-1343

# Description of changes

Implemented the new API endpoints according to the spec. 

IMPORTANT: the database structure is now quite different, so if there is already data in the existing database we need to do some migrations @janvanmansum.

Also refactored the code a bit:
* removed unneeded interfaces
* deleted the duplicated models (DD-1343), all of them now end in Entity and have 1 Repository.
* moved all logic to Usecases class instead of inside services and resources
* renamed output packages for generated code to match dropwizard conventions
* removed duplicate html view
* added proper integration tests with inmemory database for some of the usecases

I suspect there is some merge which added some refactoring added to the the old code and this went unnoticed.

# How to test

# Open issues
* The OCFL object metadata reader uses jsonpath to read the JSONLD version of the oai-ore document, but it has not yet been tested with the version that the vault ingest flow generates. It would be wiser to implement it using Apache Jena (if that does not pose problems) or at least get the namespaces sorted out correctly.
* The service was designed to be dumb, but I believe that will cause problems with data being overwritten accidentally, and generating duplicate bag/version combinations because there is no locking. 
* Root response (the / endpoint) has a version, what is this version about?

(Add links)

*

# Notify

@DANS-KNAW/dataversedans

### CODE REVIEW

File: docs/api.yml [hunk 1]
```diff
@@ -107,7 +111,7 @@ paths:
           application/json:
             schema:
               oneOf:
```
Comment (lindareijnhoudt) [inline docs/api.yml]: the oneOf is a remainder of a previous version of this api. it can be removed

File: docs/api.yml [hunk 2]
```diff
@@ -118,11 +122,15 @@ paths:
...
       responses:
         201:
           description: The object has been created
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/OcflObjectVersion'
         400:
           description: The OcflObject was syntactically incorrect, for instance not all required fields were provided.
         401:
```
Comment (lindareijnhoudt) [inline docs/api.yml:136]: the 409 conflict a few lines down is not needed if the service is not checking for conflicts

**Review by lindareijnhoudt — COMMENTED** [review]
i only looked at the api.yml document

File: src/main/java/nl/knaw/dans/catalog/core/UseCases.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            .swordToken(parameters.getSwordToken())
+            .nbn(parameters.getNbn())
+            .dataSupplier(parameters.getDataSupplier())
+            .dataversePid(parameters.getDataversePid())
+            .dataversePidVersion(parameters.getDataversePidVersion())
+            .otherId(parameters.getOtherId())
+            .otherIdVersion(parameters.getOtherIdVersion())
+            .ocflObjectPath(parameters.getOcflObjectPath())
+            .filePidToLocalPath(parameters.getFilePidToLocalPath())
+            .exportTimestamp(parameters.getExportTimestamp())
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/catalog/core/UseCases.java]: I thought the convention for these method names was `with....`

Comment (jo-pol) [reply src/main/java/nl/knaw/dans/catalog/core/UseCases.java]: Oh, apparently generated with the Builder annotation.

File: src/main/java/nl/knaw/dans/catalog/db/OcflObjectVersionEntity.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import java.time.OffsetDateTime;
+import java.util.Objects;
+
+@Entity
+@Table(name = "ocfl_object_versions")
+@Getter
+@Setter
+@ToString
+@Builder
+@AllArgsConstructor
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/catalog/db/OcflObjectVersionEntity.java]: Does the DB use this? for manually written calls I would prefer a builder.

File: src/main/java/nl/knaw/dans/catalog/db/OcflObjectVersionEntityRepository.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    @Override
+    public List<OcflObjectVersionEntity> findAll() {
+        return new ArrayList<>(list(query("from OcflObjectVersionEntity")));
+    }
+
+    @Override
+    public List<OcflObjectVersionEntity> findAllBySwordToken(String swordToken) {
+        return new ArrayList<>(
+            query("from OcflObjectVersionEntity where swordToken = :swordToken order by objectVersion desc limit 1")
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/catalog/db/OcflObjectVersionEntityRepository.java]: findall yet retrieving just one? I suppose I don't understand.

File: src/main/java/nl/knaw/dans/catalog/db/TarEntity.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    @OneToMany(cascade = {CascadeType.ALL}, mappedBy = "tar")
+    @ToString.Exclude
+    private List<OcflObjectVersionEntity> ocflObjectVersions = new ArrayList<>();
+
+    public List<TarPartEntity> getTarParts() {
+        return new ArrayList<>(tarParts);
+    }
+
+    public void setTarParts(List<TarPartEntity> tarParts) {
+        if (this.tarParts == null) {
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/catalog/db/TarEntity.java]: Thought I saw some clever call for this somewhere.

File: src/main/java/nl/knaw/dans/catalog/db/TarPartEntity.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    @Column(name = "checksum_algorithm")
+    private String checksumAlgorithm;
+    @Column(name = "checksum_value")
+    private String checksumValue;
+    @ManyToOne
+    @JoinColumn(name = "tar_uuid")
+    private TarEntity tar;
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/catalog/db/TarPartEntity.java]: No annotation for this?

File: src/main/resources/nl/knaw/dans/catalog/resource/view/error.ftl [hunk 8]
```diff
@@ -16,15 +16,15 @@
                     <p><em>${errorMessage.message}</em></p>
                     <p>There was an error displaying this page.</p>
                 </header>
+
                 <#if errorMessage.details?has_content>
-                <div class="bg-light my-5 py-4 px-4 border">
-                    <pre class="my-0"><code>${errorMessage.details}</code></pre>
-                </div>
-               </#if>
+                    <div class="bg-light my-5 py-4 px-4 border">
```
Comment (jo-pol) [inline src/main/resources/nl/knaw/dans/catalog/resource/view/error.ftl:21]: ftl is new to me

**Review by jo-pol — COMMENTED** [review]
just some dumb question/remarks, skipped src/test

File: src/main/java/nl/knaw/dans/catalog/db/TarEntity.java (hunk 6 above)
Comment (jo-pol) [reply src/main/java/nl/knaw/dans/catalog/db/TarEntity.java]: The memory resurfaces. It was about adding a(nother) value to a Map<String,List<...>>. That is another situation.

File: src/main/java/nl/knaw/dans/catalog/core/UseCases.java (hunk 3 above)
Comment (ericdevries) [reply src/main/java/nl/knaw/dans/catalog/core/UseCases.java]: These are indeed openapi generated classes

File: src/main/java/nl/knaw/dans/catalog/db/OcflObjectVersionEntity.java (hunk 4 above)
Comment (ericdevries) [reply src/main/java/nl/knaw/dans/catalog/db/OcflObjectVersionEntity.java]: Yes the builder uses it; however we can set it to package-private to not allow access to it from the outside

File: src/main/java/nl/knaw/dans/catalog/db/OcflObjectVersionEntityRepository.java (hunk 5 above)
Comment (ericdevries) [reply src/main/java/nl/knaw/dans/catalog/db/OcflObjectVersionEntityRepository.java]: Good point, some leftovers I forgot to delete

File: src/main/java/nl/knaw/dans/catalog/db/TarPartEntity.java (hunk 7 above)
Comment (ericdevries) [reply src/main/java/nl/knaw/dans/catalog/db/TarPartEntity.java]: Yes, but they recommend against doing this for JPA entities. See https://dzone.com/articles/lombok-and-jpa-what-may-go-wrong for more information.

File: src/main/resources/nl/knaw/dans/catalog/resource/view/error.ftl (hunk 8 above)
Comment (ericdevries) [reply src/main/resources/nl/knaw/dans/catalog/resource/view/error.ftl:21]: It is a freemarker template, which allows for text to be interpolated with values. In this case it is a html document.

Comment (ericdevries) [reply src/main/resources/nl/knaw/dans/catalog/resource/view/error.ftl:21]: See https://freemarker.apache.org/ for more info

File: src/main/java/nl/knaw/dans/catalog/db/TarPartEntity.java (hunk 7 above)
Comment (jo-pol) [reply src/main/java/nl/knaw/dans/catalog/db/TarPartEntity.java]: I would be tempted to change in ignorance. Comment everywhere might also be overkill. Don't know the balance.

**Review by jo-pol — DISMISSED** [review]
Don't want to block this work in progress.

File: src/main/java/nl/knaw/dans/catalog/db/TarPartEntity.java (hunk 7 above)
Comment (ericdevries) [reply src/main/java/nl/knaw/dans/catalog/db/TarPartEntity.java]: I have a plugin for IntelliJ by the creators of that article that warns me about that. It should also be possible to create some kind of linting rule that detects both @Entity and @Data on a class and rejects it.

**Review by janvanmansum — APPROVED** [review]

**Pull request merged**